stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/ChannelControllerImpl.kt

```diff
@@ -608,26 +630,43 @@ internal class ChannelControllerImpl(
 
         return if (online) {
             // upload attachments
-            logger.logI("Uploading attachments for message with id ${newMessage.id} and text ${newMessage.text}")
-            newMessage.attachments = newMessage.attachments.map {
-                var attachment: Attachment = it
-                if (it.upload != null) {
-                    val result = uploadAttachment(it, attachmentTransformer)
-                    attachment.uploadComplete = result.isSuccess
-                    if (result.isSuccess) {
-                        attachment = result.data()
-                    } else {
-                        attachment.uploadState = Attachment.UploadState.Failed(result.error())
-                        logger.logE("Failed to upload attachment for message")
+
+            if (hasAttachments) {
+                logger.logI("Uploading attachments for message with id ${newMessage.id} and text ${newMessage.text}")
+
+                newMessage.attachments = newMessage.attachments.mapIndexed { i, attach ->
```

Comment:
Why do you need local `attachmentProgressList ` and mapping with index? ProgressTrackerFactory already can return existing progressTracker by some id.

Reply:
It is possible to create the ProgressTracker inside the `map` and don't use the index, but I create `attachmentProgressList` before `upsertMessage(newMessage)` is called, so when the `FileAttachmentsView::bind` is called all the `ProgressTracker` are already created. This makes the debug process easier... otherwise it would be the adapter creating the trackers.

Reply:
I create all the trackers and store then in `attachmentProgressList` before the `upsertMessage(newMessage)` is called, so when the `FileAttachmentViewHolder::bind` is called all the trackers are already created. 

Although managing the instances with the factory would work, the adapter would create the trackers first if I call the `getOrCreate` inside `newMessage.attachments.mapIndexed`, making the logic harder to reason about. I could create the trackers and not assigning then to any list/variable, but that would look odd to me too.